sigstore/sign.py

```diff
@@ -192,6 +231,75 @@ def sign(
         )
 
 
+class SigningContext:
+    """
+    Keep a context between signing operations.
+    """
+
+    def __init__(
+        self,
+        *,
+        fulcio: FulcioClient,
+        rekor: RekorClient,
+    ):
+        """
+        Create a new `SigningContext`.
+
+        `fulcio` is a `FulcioClient` capable of connecting to a Fulcio instance
+        and returning signing certificates.
+
+        `rekor` is a `RekorClient` capable of connecting to a Rekor instance
+        and creating transparency log entries.
+        """
+        self._fulcio = fulcio
+        self._rekor = rekor
+
+    @classmethod
+    def production(cls) -> SigningContext:
+        """
+        Return a `SigningContext` instance configured against Sigstore's production-level services.
+        """
+        updater = TrustUpdater.production()
+        rekor = RekorClient.production(updater)
+        return cls(
+            fulcio=FulcioClient.production(),
+            rekor=rekor,
+        )
+
+    @classmethod
+    def staging(cls) -> SigningContext:
+        """
+        Return a `SignerContext` instance configured against Sigstore's staging-level services.
+        """
+        updater = TrustUpdater.staging()
+        rekor = RekorClient.staging(updater)
+        return cls(
+            fulcio=FulcioClient.staging(),
+            rekor=rekor,
+        )
+
+    @contextmanager
+    def with_signer(
+        self, identity_token: str, cache: bool = True
+    ) -> Generator[Signer, None, None]:
+        """
+        A context manager for signing operations.
+
+        `identity_token` is the identity token passed to the `Signer` instance
+        and used to request a signing certificate from Fulcio.
+
+        `cache` determines whether the signing certificate and ephemeral private key
+        generated by the `Signer` instance should be reused (until the certificate expires)
+        to sign different artifacts.
+        Default is `True`.
+        """
+        signer = Signer(identity_token, cache)
+        try:
+            yield signer
+        finally:
+            del signer
```

Comment:
I've got mixed feelings on `del` in Python 3: it can lead people into a false sense of security/"finality" that isn't actually guaranteed (e.g. anything taking a reference on any `Signer` instance state will prevent a full object deletion). 

OTOH, there's nothing unsafe about it either -- if we _do_ keep this, we might as well go 100% and also define `Signer.__del__` to include explicit destruction of the secret key and other materials. 

CC @jku for opinions.

Reply:
Or another idea: rather than `del`, we could create a "poison" style API, e.g.:

```
signer._poison()

# any attempts to sign or use the signer object after this point raise an `InvalidObjectError` or similar
```

This has the benefit of correctly "enforcing" the lifetime of the `signer` binding in a way that context managers can't, at the cost of a decent bit of complexity (which I'm not sure it's worth).

Reply:
The first approach with defining a `__del__` method to get rid of the signer's attributes seems like a simpler option to me. Even if this does not happen with `del`, it makes sense that the attributes get deleted when leaving the signing context scope.

Reply:
This _might_ be a little unexpected for contextmanager users: I think usually the context object ( signer in this case) still exists after the with-clause, it's just closed or inactive. In this case I think the whole object just disappears  ... 

I think it's fine though. Maybe worth a docstring mention?

Reply:
Yeah, that's a good point -- I did a little more reading, and it seems like defining `__del__` isn't considered idiomatic in Python 3 anymore (both because of end user expectations, and because it's misleading -- it doesn't actually guarantee that the secret materials here have been freed).

Given that, I think we can drop `del` and `__del__` entirely and emphasize in documentation that the `Signer` object is only valid in its created context.